<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

# Lab: Cleaning Rock Song Data

---


In [381]:
import pandas as pd
import numpy as np 
import seaborn as sns

%matplotlib inline

### 1. Load `rock.csv` and do an initial examination of its data columns.

In [382]:
rockfile = "../../../../resource-datasets/rock_songs/rock.csv"

In [383]:
# Load the data.
df=pd.read_csv(rockfile)

In [484]:
# Look at the information regarding its columns.
df.head()

song       artist release  \
0        Caught Up in You  .38 Special    1982   
1            Fantasy Girl  .38 Special       0   
2         Hold On Loosely  .38 Special    1981   
3  Rockin' Into the Night  .38 Special    1980   
4       Art For Arts Sake         10cc    1975   

                             song+artist  first  year  playcount  f*g  \
0        Caught Up in You by .38 Special      1     1         82   82   
1            Fantasy Girl by .38 Special      1     0          3    0   
2         Hold On Loosely by .38 Special      1     1         85   85   
3  Rockin' Into the Night by .38 Special      1     1         18   18   
4              Art For Arts Sake by 10cc      1     1          1    1   

   higher_than_1970  
0              True  
1             False  
2              True  
3              True  
4              True

### 2.  Clean up the column names.

Let's clean up the column names. There are two ways we can accomplish this:

#### 2.A Change the column names when you import the data using `pd.read_csv()`.

Notice that, when passing `names=[..A LIST OF STRING..]` with a number of columns that matches the number of strings in the passed list, you replace the column names.

NOTE: When you create custom column names, the first row of the `.csv` already represents a header. It is important to tell `pandas` to skip that row. The `skiprows=1` keyword argument to `read_csv()` will tell `pandas` to skip the first row.

In [385]:
# Change the column names when loading the '.csv':
df=pd.read_csv("../../../../resource-datasets/rock_songs/rock.csv",names=['song','artist','released',
                                                                          'song+artist','first','year',
                                                                          'playcount','fg'],skiprows=1)
df.head()

song       artist released  \
0        Caught Up in You  .38 Special     1982   
1            Fantasy Girl  .38 Special      NaN   
2         Hold On Loosely  .38 Special     1981   
3  Rockin' Into the Night  .38 Special     1980   
4       Art For Arts Sake         10cc     1975   

                             song+artist  first  year  playcount  fg  
0        Caught Up in You by .38 Special      1     1         82  82  
1            Fantasy Girl by .38 Special      1     0          3   0  
2         Hold On Loosely by .38 Special      1     1         85  85  
3  Rockin' Into the Night by .38 Special      1     1         18  18  
4              Art For Arts Sake by 10cc      1     1          1   1

#### 2.B Change column names using the `.rename()` function.

The `.rename()` function takes an argument, `columns=name_dict`, in which `name_dict` is a dictionary containing the original column names as keys and the new column names as values.

In [386]:
# Change the column names using the `.rename()` function.
new_dict={'song':'songs','artist':'arti','released':'release',
                   'song+artist':'songs+arti','first':'first?','year':'year?','playcount':'playc','fg':'f*g'}
df.rename(columns=new_dict,inplace=True)
df.head()

songs         arti release  \
0        Caught Up in You  .38 Special    1982   
1            Fantasy Girl  .38 Special     NaN   
2         Hold On Loosely  .38 Special    1981   
3  Rockin' Into the Night  .38 Special    1980   
4       Art For Arts Sake         10cc    1975   

                              songs+arti  first?  year?  playc  f*g  
0        Caught Up in You by .38 Special       1      1     82   82  
1            Fantasy Girl by .38 Special       1      0      3    0  
2         Hold On Loosely by .38 Special       1      1     85   85  
3  Rockin' Into the Night by .38 Special       1      1     18   18  
4              Art For Arts Sake by 10cc       1      1      1    1

#### 2.C Reassigning the `.columns` attribute of a DataFrame.

You can also just reassign the `.columns` attribute to a list of strings containing the new column names. 

The only caveat with reassigning `.columns` is that you have to reassign all of the column names at once. You can't partially replace a value by working on `.columns` directly. You have to reassign `.columns` with a list of equal length. 

In [387]:
# Replace the column names by reassigning the `.columns` attribute.
df.columns=['song','artist','release','song+artist','first','year','playcount','f*g']
df.head()

song       artist release  \
0        Caught Up in You  .38 Special    1982   
1            Fantasy Girl  .38 Special     NaN   
2         Hold On Loosely  .38 Special    1981   
3  Rockin' Into the Night  .38 Special    1980   
4       Art For Arts Sake         10cc    1975   

                             song+artist  first  year  playcount  f*g  
0        Caught Up in You by .38 Special      1     1         82   82  
1            Fantasy Girl by .38 Special      1     0          3    0  
2         Hold On Loosely by .38 Special      1     1         85   85  
3  Rockin' Into the Night by .38 Special      1     1         18   18  
4              Art For Arts Sake by 10cc      1     1          1    1

### 3. Subsetting data where null values exist.

We have mixed `str` and `NaN` values in the `release` column. `NaN` stands for "not a number" and is the way `pandas` handles "nulls" or nonexistent data. We can use the `.isnull()` method of a Series to find null values.

Print the header of the data subset to where the `release` column is null values.

In [388]:
# Show records where df['release'] is null
df[df['release'].isnull()].head()

song       artist release  \
1             Fantasy Girl  .38 Special     NaN   
10   Baby, Please Don't Go        AC/DC     NaN   
13  CAN'T STOP ROCK'N'ROLL        AC/DC     NaN   
16        Girls Got Rhythm        AC/DC     NaN   
24         Let's Get It Up        AC/DC     NaN   

                        song+artist  first  year  playcount  f*g  
1       Fantasy Girl by .38 Special      1     0          3    0  
10   Baby, Please Don't Go by AC/DC      1     0          1    0  
13  CAN'T STOP ROCK'N'ROLL by AC/DC      1     0          5    0  
16        Girls Got Rhythm by AC/DC      1     0         24    0  
24         Let's Get It Up by AC/DC      1     0          4    0

### 4. Update slices of your DataFrame based on mask selection/slices.

In many scenarios, we want to upate values in our DataFrame according to criteria. Let's say we wanted to set all of the null values in `release` to 0.

With newer versions of `pandas`, in order to manipulate data in the original DataFrame, we have to use `.loc` while performing reassignment using a mask and an index.

For example, the following won't always work:
```python
df[row_mask]['column_name'] = new_value
```

The best way to accomplish the same task is:
```python
df.loc[row_mask, 'column_name'] = new_value
```

For multiple column assignment, you would use:
```python
df.loc[row_mask, ['col_1', 'col_2', 'col_3']] = new_value
```

#### 4.A Let's try it out. Make all of the null values in `release` 0.

In [389]:
# Replace release nulls with 0
df.loc[df[df['release'].isnull()].index,'release']=0
df.head(15)

song         artist release  \
0              Caught Up in You    .38 Special    1982   
1                  Fantasy Girl    .38 Special       0   
2               Hold On Loosely    .38 Special    1981   
3        Rockin' Into the Night    .38 Special    1980   
4             Art For Arts Sake           10cc    1975   
5                    Kryptonite   3 Doors Down    2000   
6                         Loser   3 Doors Down    2000   
7                 When I'm Gone   3 Doors Down    2002   
8                    What's Up?  4 Non Blondes    1992   
9                    Take On Me           a-ha    1985   
10        Baby, Please Don't Go          AC/DC       0   
11                Back In Black          AC/DC    1980   
12                      Big Gun          AC/DC    1993   
13       CAN'T STOP ROCK'N'ROLL          AC/DC       0   
14  Dirty Deeds Done Dirt Cheap          AC/DC    1976   

                              song+artist  first  year  playcount  f*g  
0         Caught Up in You by .38 Special      1     1         82   82  
1             Fantasy Girl by .38 Special      1     0          3    0  
2          Hold On Loosely by .38 Special      1     1         85   85  
3   Rockin' Into the Night by .38 Special      1     1         18   18  
4               Art For Arts Sake by 10cc      1     1          1    1  
5              Kryptonite by 3 Doors Down      1     1         13   13  
6                   Loser by 3 Doors Down      1     1          1    1  
7           When I'm Gone by 3 Doors Down      1     1          6    6  
8             What's Up? by 4 Non Blondes      1     1          3    3  
9                      Take On Me by a-ha      1     1          1    1  
10         Baby, Please Don't Go by AC/DC      1     0          1    0  
11                 Back In Black by AC/DC      1     1         97   97  
12                       Big Gun by AC/DC      1     1          6    6  
13        CAN'T STOP ROCK'N'ROLL by AC/DC      1     0          5    0  
14   Dirty Deeds Done Dirt Cheap by AC/DC      1     1         85   85

#### 4.B Verify that `release` contains no null values.

In [390]:
# A:
df[df['release'].isnull()]

Empty DataFrame
Columns: [song, artist, release, song+artist, first, year, playcount, f*g]
Index: []

### 5. Ensure that the data types of the columns make sense. 

Verifying column data types is a critical part of data munging. If columns have the wrong data type, then there is usually corrupted or incorrect data in some of the observations.

#### 5.A Look at the data types for the columns. Are any incorrect given what the data represents?

In [391]:
# A:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2230 entries, 0 to 2229
Data columns (total 8 columns):
song           2230 non-null object
artist         2230 non-null object
release        2230 non-null object
song+artist    2230 non-null object
first          2230 non-null int64
year           2230 non-null int64
playcount      2230 non-null int64
f*g            2230 non-null int64
dtypes: int64(4), object(4)
memory usage: 139.5+ KB


### 6. Investigate and clean up the `release` column.

The `release` column is a string data type when it should be an integer.

#### 6.A Figure out what value(s) are causing the `release` column to be encoded as a string instead of an integer.

In [392]:
# A:
df[(df['release']!=np.int()) & (df['release']!=np.str())].head()

song        artist release  \
0        Caught Up in You   .38 Special    1982   
2         Hold On Loosely   .38 Special    1981   
3  Rockin' Into the Night   .38 Special    1980   
4       Art For Arts Sake          10cc    1975   
5              Kryptonite  3 Doors Down    2000   

                             song+artist  first  year  playcount  f*g  
0        Caught Up in You by .38 Special      1     1         82   82  
2         Hold On Loosely by .38 Special      1     1         85   85  
3  Rockin' Into the Night by .38 Special      1     1         18   18  
4              Art For Arts Sake by 10cc      1     1          1    1  
5             Kryptonite by 3 Doors Down      1     1         13   13

#### 6.B Look at the rows in which there is incorrect data in the `release` column.

In [393]:
# A:
for i in range(len(df)):
    try:
        float(df['release'][i])
    except:       
        print(df.iloc[i])
        print('row number: ',i)

song                             Bullfrog Blues
artist                           Rory Gallagher
release                           SONGFACTS.COM
song+artist    Bullfrog Blues by Rory Gallagher
first                                         1
year                                          1
playcount                                     1
f*g                                           1
Name: 1504, dtype: object
row number:  1504


#### 6.C. Clean up the data. Normally we may replace the offending data with null np.nan values, however we previously converted all of the nan values in the release column to zeros so we might as well continue with the same practice. Replacing with 0 (or nan) will allow us to convert the column to numeric.

In [394]:
# A:
k=0
for i in range(len(df)):
    try:
        float(df['release'][i])
    except:       
        df.loc[i,'release']=0
        k=i
df.loc[:,'release'].astype('int64')
df.loc[k]

song                             Bullfrog Blues
artist                           Rory Gallagher
release                                       0
song+artist    Bullfrog Blues by Rory Gallagher
first                                         1
year                                          1
playcount                                     1
f*g                                           1
Name: 1504, dtype: object

### 7. Get summary statistics for the `release` column using the `.describe()` function.

Now that the `release` column is finally a numeric data type, we can apply the `.describe()` function.  

#### 7.A Print out the summary stats for the `release` column. What is the earliest and latest release date?

In [406]:
# A:
df['release'].describe()

count     2230
unique      56
top          0
freq       578
Name: release, dtype: int64

#### 7.B Based on the summary statistics, is there anything else wrong with the `release` column? 

# A: yes many of it is zeros, with 578 zeros.

### 8. Make changes and investigate using custom functions with `.apply()`.

Let's say we want to traverse every single row in our data set and apply a function to that row.

#### 8.A Write a function that will take a row of a DataFrame and print out the song, artist, and whether or not the release date is < 1970.

In [480]:
# A:
def date_higher_1970(x):
    return np.int_(x['release'])>1970

#### 8.B Using the `.apply()` function, apply the function you wrote to the first four rows of the DataFrame.

You will need to tell the `apply` function to operate row by row. Setting the keyword argument as `axis=1` indicates that the function should be applied to each row individually.

In [482]:
# A:
df['higher_than_1970']=df.apply(date_higher_1970,axis=1)
df[['song','artist','higher_than_1970']]

song         artist  \
0                                Caught Up in You    .38 Special   
1                                    Fantasy Girl    .38 Special   
2                                 Hold On Loosely    .38 Special   
3                          Rockin' Into the Night    .38 Special   
4                               Art For Arts Sake           10cc   
5                                      Kryptonite   3 Doors Down   
6                                           Loser   3 Doors Down   
7                                   When I'm Gone   3 Doors Down   
8                                      What's Up?  4 Non Blondes   
9                                      Take On Me           a-ha   
10                          Baby, Please Don't Go          AC/DC   
11                                  Back In Black          AC/DC   
12                                        Big Gun          AC/DC   
13                         CAN'T STOP ROCK'N'ROLL          AC/DC   
14                    Dirty Deeds Done Dirt Cheap          AC/DC   
15                        For Those About To Rock          AC/DC   
16                               Girls Got Rhythm          AC/DC   
17                                 Hard As A Rock          AC/DC   
18                             Have a Drink On Me          AC/DC   
19                                    Hells Bells          AC/DC   
20                                Highway To Hell          AC/DC   
21                     It's A Long Way To The Top          AC/DC   
22                                      Jailbreak          AC/DC   
23                              Let There Be Rock          AC/DC   
24                                Let's Get It Up          AC/DC   
25                                      Live Wire          AC/DC   
26                                     Moneytalks          AC/DC   
27                                  Night Prowler          AC/DC   
28            Rock and Roll Ain't Noise Pollution          AC/DC   
29                                Shoot To Thrill          AC/DC   
...                                           ...            ...   
2200                      Owner of a Lonely Heart            Yes   
2201                                   Roundabout            Yes   
2202                             Starship Trooper            Yes   
2203                            Wonderous Stories            Yes   
2204                         Yours Is No Disgrace            Yes   
2205                        Tell Me What You Want          Zebra   
2206                        Who's Behind The Door          Zebra   
2207                           Time of the Season        Zombies   
2208                    A Fool for Your Stockings         ZZ Top   
2209             Arrested For Driving While Blind         ZZ Top   
2210                 Beer Drinkers & Hell Raisers         ZZ Top   
2211                             Cheap Sunglasses         ZZ Top   
2212                         Gimme All Your Lovin         ZZ Top   
2213                        Got Me Under Pressure         ZZ Top   
2214                            Heard It On the X         ZZ Top   
2215                                  I Thank You         ZZ Top   
2216                      I'm Bad, I'm Nationwide         ZZ Top   
2217                      Jesus Just Left Chicago         ZZ Top   
2218                                Just Got Paid         ZZ Top   
2219                                    La Grange         ZZ Top   
2220                                         Legs         ZZ Top   
2221                     My Head's In Mississippi         ZZ Top   
2222                           Party On The Patio         ZZ Top   
2223                               Pearl Necklace         ZZ Top   
2224                            Sharp Dressed Man         ZZ Top   
2225                      She Loves My Automobile         ZZ Top   
2226                            Tube Snake Boogie         ZZ Top   
2227                             

You'll notice that there will be a final output Series of `None` values. The `.apply()` function, if a return value is not specified, will return a Series of `None` values (similar to how the default return for Python functions is `None` when a return statement is not specified).

### 9. Write a function that converts cells in a DataFrame to float and otherwise replaces them with `np.nan`.

If applied to our data, it would keep only the numeric information and otherwise input null values.

Recall that the try-except syntax in Python is a great way to try something and take another action if the initial step fails:

```python
try:
    Perform some action.
except:
   Perform some other action if the first failed with an error.
```

#### 9.A Write the function that takes a column and converts all of its values to float if possible and `np.nan` otherwise. The return value should be the converted Series.

In [491]:
# A:
def float_or_nan(x):
    try:
        return float(x)
    except:
        return np.nan

#### 9.B Try your function out on the rock song data and ensure the output is what you expected.


In [493]:
# A:
df.applymap(float_or_nan).head()

song  artist  release  song+artist  first  year  playcount   f*g  \
0   NaN     NaN   1982.0          NaN    1.0   1.0       82.0  82.0   
1   NaN     NaN      0.0          NaN    1.0   0.0        3.0   0.0   
2   NaN     NaN   1981.0          NaN    1.0   1.0       85.0  85.0   
3   NaN     NaN   1980.0          NaN    1.0   1.0       18.0  18.0   
4   NaN     NaN   1975.0          NaN    1.0   1.0        1.0   1.0   

   higher_than_1970  
0               1.0  
1               0.0  
2               1.0  
3               1.0  
4               1.0

#### 9.C Describe the new float-only DataFrame.

In [498]:
# A:
df.applymap(float_or_nan).describe()

song  artist      release  song+artist   first         year  \
count     2.000000     0.0  2230.000000          0.0  2230.0  2230.000000   
mean   1012.000000     NaN  1465.331390          NaN     1.0     0.741256   
std    1367.544515     NaN   867.196161          NaN     0.0     0.438043   
min      45.000000     NaN     0.000000          NaN     1.0     0.000000   
25%     528.500000     NaN     0.000000          NaN     1.0     0.000000   
50%    1012.000000     NaN  1973.000000          NaN     1.0     1.000000   
75%    1495.500000     NaN  1981.000000          NaN     1.0     1.000000   
max    1979.000000     NaN  2014.000000          NaN     1.0     1.000000   

         playcount          f*g  higher_than_1970  
count  2230.000000  2230.000000       2230.000000  
mean     16.872646    15.048430          0.584753  
std      25.302972    25.288366          0.492875  
min       0.000000     0.000000          0.000000  
25%       1.000000     0.000000          0.000000  
50%       4.000000     3.000000          1.000000  
75%      21.000000    18.000000          1.000000  
max     142.000000   142.000000          1.000000

### 10. What are the top 20 most popular songs by plays?

In [499]:
# A:
df.sort_values(by='playcount',ascending=False).head(20)

song             artist release  \
49                       Dream On          Aerosmith    1973   
868      All Along the Watchtower       Jimi Hendrix    1968   
65                  Sweet Emotion          Aerosmith    1975   
38    You Shook Me All Night Long              AC/DC    1980   
267           More Than a Feeling             Boston    1976   
968          Carry On Wayward Son             Kansas    1976   
269                 Peace of Mind             Boston    1976   
774                  Crazy On You              Heart    1976   
2220                         Legs             ZZ Top    1983   
2224            Sharp Dressed Man             ZZ Top    1983   
989        Rock And Roll All Nite               Kiss    1975   
1354            Bohemian Rhapsody              Queen    1975   
1675                     Renegade               Styx    1978   
784                     Magic Man              Heart    1976   
956              Wheel in the Sky            Journey    1978   
1106         Blinded by the Light       Manfred Mann    1976   
1621                    The Joker  Steve Miller Band    1973   
1523                   Tom Sawyer               Rush    1981   
1866           Just What I Needed           The Cars    1978   
772                     Barracuda              Heart    1977   

                                   song+artist  first  year  playcount  f*g  \
49                       Dream On by Aerosmith      1     1        142  142   
868   All Along the Watchtower by Jimi Hendrix      1     1        141  141   
65                  Sweet Emotion by Aerosmith      1     1        141  141   
38        You Shook Me All Night Long by AC/DC      1     1        138  138   
267              More Than a Feeling by Boston      1     1        134  134   
968             Carry On Wayward Son by Kansas      1     1        134  134   
269                    Peace of Mind by Boston      1     1        132  132   
774                      Crazy On You by Heart      1     1        125  125   
2220                            Legs by ZZ Top      1     1        121  121   
2224               Sharp Dressed Man by ZZ Top      1     1        120  120   
989             Rock And Roll All Nite by Kiss      1     1        119  119   
1354                Bohemian Rhapsody by Queen      1     1        119  119   
1675                          Renegade by Styx      1     1        116  116   
784                         Magic Man by Heart      1     1        115  115   
956                Wheel in the Sky by Journey      1     1        114  114   
1106      Blinded by the Light by Manfred Mann      1     1        114  114   
1621            The Joker by Steve Miller Band      1     1        114  114   
1523                        Tom Sawyer by Rush      1     1        114  114   
1866            Just What I Needed by The Cars      1     1        113  113   
772                         Barracuda by Heart      1     1        113  113   

      higher_than_1970  
49                True  
868              False  
65                True  
38                True  
267               True  
968               True  
269               True  
774               True  
2220              True  
2224              True  
989               True  
1354              True  
1675              True  
784               True  
956               True  
1106              True  
1621              True  
1523              True  
1866              True  
772               True

### 11. Which years have the most plays?

In [515]:
# A:
df.sort_values(by='playcount',ascending=False)[['release','playcount']].head()

release  playcount
49     1973        142
868    1968        141
65     1975        141
38     1980        138
267    1976        134

### 12. Which records don't have matching "Play Count" corresponding to "F*G"?

In [530]:
# A:
df[df['release'] == df['f*g']]

song                    artist release  \
1                           Fantasy Girl               .38 Special       0   
10                 Baby, Please Don't Go                     AC/DC       0   
13                CAN'T STOP ROCK'N'ROLL                     AC/DC       0   
16                      Girls Got Rhythm                     AC/DC       0   
24                       Let's Get It Up                     AC/DC       0   
25                             Live Wire                     AC/DC       0   
26                            Moneytalks                     AC/DC       0   
29                       Shoot To Thrill                     AC/DC       0   
31                              Sin City                     AC/DC       0   
35        What Do You Do For Money Honey                     AC/DC       0   
42                 Baby, Please Don't Go                 Aerosmith       0   
45                         Come Together                 Aerosmith       0   
53                      Kings and Queens                 Aerosmith       0   
60                        One Way Street                 Aerosmith       0   
66                        The Other Side                 Aerosmith       0   
71                     You See Me Crying                 Aerosmith       0   
75                               Fantasy                 Aldo Nova       0   
76                           BE MY LOVER              Alice Cooper       0   
80                         House Of Fire              Alice Cooper       0   
100                          One Way Out      Allman Brothers Band       0   
102                           Southbound      Allman Brothers Band       0   
104                        Whippin' Post      Allman Brothers Band       0   
110                       Immigrant Song                Ann Wilson       0   
111                               Roller                April Wine       0   
113              Coming into Los Angeles              Arlo Guthrie       0   
114                              Wayside             Artful Dodger       0   
118                               Spooky    Atlanta Rhythm Section       0   
120                Blondes In Black Cars                 Autograph       0   
122   Rock And Roll Party In The Streets                       Axe       0   
124             Roll On Down The Highway  Bachman-Turner Overdrive       0   
...                                  ...                       ...     ...   
2149              Take Your Whiskey Home                 Van Halen       0   
2150                           Top Jimmy                 Van Halen       0   
2151                    Top of the World                 Van Halen       0   
2157                    And It Stoned Me              Van Morrison       0   
2158                          BLUE MONEY              Van Morrison       0   
2162                              Gloria              Van Morrison       0   
2164                  Jackie Wilson Said              Van Morrison       0   
2168              Edge of a Broken Heart                     Vixen       0   
2174                   Uncle Tom's Cabin                   Warrant       0   
2178              Street Corner Serenade                Wet Willie       0   
2179                                Wait                White Lion       0   
2184                       Slow An' Easy                Whitesnake       0   
2185                  Still of the Night                Whitesnake       0   
2186                        Who Are You?                   Why Not       0   
2188                         Night Moves                  Wildlife       0   
2189                   We've Got Tonight                  Wildlife       0   
2190                    Can't Get Enough                    Winger       0   
2191             Headed For A Heartbreak                    Winger       0   
2192                           Seventeen                    Winger       0   
2194                    Summertime Girls                  

### Bonus: Which artists have the most missing values between each of the variables? 

In [544]:
# A:
